File: entwurfsmuster-beispiele/ruby/builder.rb
# Product: Pizza
class Pizza
    attr_accessor :crust, :sauce, :cheese, :toppings
  
    def initialize
      @crust = nil
      @sauce = nil
      @cheese = nil
      @toppings = []
    end
  
    def describe
      puts "Pizza: #{crust}, #{sauce}, #{cheese}, Toppings: #{toppings.join(', ')}"
    end
  end
  
  # Builder Interface
  class PizzaBuilder
    def build_crust
      raise NotImplementedError, "#{self.class} must implement build_crust"
    end
  
    def build_sauce
      raise NotImplementedError, "#{self.class} must implement build_sauce"
    end
  
    def build_cheese
      raise NotImplementedError, "#{self.class} must implement build_cheese"
    end
  
    def build_toppings
      raise NotImplementedError, "#{self.class} must implement build_toppings"
    end
  
    def pizza
      raise NotImplementedError, "#{self.class} must implement pizza"
    end
  end
  
  # Concrete Builder: MargheritaPizzaBuilder
  class MargheritaPizzaBuilder < PizzaBuilder
    def build_crust
      pizza.crust = "Thin Crust"
    end
  
    def build_sauce
      pizza.sauce = "Tomato Sauce"
    end
  
    def build_cheese
      pizza.cheese = "Mozzarella Cheese"
    end
  
    def build_toppings
      pizza.toppings = ["Tomatoes", "Basil"]
    end
  
    def pizza
      @pizza ||= Pizza.new
    end
  end
  
  # Director: PizzaDirector
  class PizzaDirector
    attr_reader :builder
  
    def initialize(builder)
      @builder = builder
    end
  
    def construct
      builder.build_crust
      builder.build_sauce
      builder.build_cheese
      builder.build_toppings
    end
  end
  
  # Client
  margherita_builder = MargheritaPizzaBuilder.new
  pizza_director = PizzaDirector.new(margherita_builder)
  pizza_director.construct
  margherita_pizza = margherita_builder.pizza
  margherita_pizza.describe
  

File: entwurfsmuster-beispiele/python/builder.py
# Product
class Pizza:
    def __init__(self):
        self.size = ""
        self.cheese = False
        self.pepperoni = False
        self.mushrooms = False

    def __str__(self):
        return f"Pizza: Size({self.size}), Cheese({self.cheese}), Pepperoni({self.pepperoni}), Mushrooms({self.mushrooms})"


# Abstract Builder
class PizzaBuilder:
    def __init__(self):
        self.pizza = Pizza()

    def build_size(self):
        pass

    def build_cheese(self):
        pass

    def build_pepperoni(self):
        pass

    def build_mushrooms(self):
        pass

    def get_result(self):
        return self.pizza


# Concrete Builder
class MargheritaPizzaBuilder(PizzaBuilder):
    def build_size(self):
        self.pizza.size = "Medium"

    def build_cheese(self):
        self.pizza.cheese = True

    def build_pepperoni(self):
        pass

    def build_mushrooms(self):
        pass


# Director
class PizzaDirector:
    def __init__(self, pizza_builder):
        self.pizza_builder = pizza_builder

    def construct_pizza(self):
        self.pizza_builder.build_size()
        self.pizza_builder.build_cheese()
        self.pizza_builder.build_pepperoni()
        self.pizza_builder.build_mushrooms()


# Client code
if __name__ == "__main__":
    margherita_builder = MargheritaPizzaBuilder()
    director = PizzaDirector(margherita_builder)

    director.construct_pizza()
    margherita_pizza = margherita_builder.get_result()

    print(margherita_pizza)
